### DISCUSSION TITLE
How to make Next.js's recommendations for <Script> align with GTM recomendations?

### DISCUSSION THREAD

**Discussion by ATK3Shape** [vercel/next.js#48011] [Help] [answered]
### Summary

**Google Tag Manager documentation says:** "The first code block is best placed immediately after the web page's opening `<head>` tag, or as high in the `<head>` as possible. This helps to ensure that your Tag Manager configuration is available and ready when the rest of the page loads." or "Add the following code so that it is as close to the opening `<head>` tag as possible on every page of your website". ([link to docs](https://developers.google.com/tag-platform/tag-manager/web))

![image](https://user-images.githubusercontent.com/24272528/230342331-4e45a6fa-b034-4af1-a1f3-c18a84f761bc.png)

**On the other hand**, in the official example ([here](https://github.com/vercel/next.js/blob/canary/examples/with-google-tag-manager/pages/_app.js)) `<Script>` from 'next/script' is used with '_afterInteractive_' strategy for adding GTM script.

**The problem**
So, GTM requires to put its code inside `<head>` tag,
but using strategy="afterInteractive" puts the script inside `<body>` tag.

**The question**
Why it is recommended way to add GTM/Analytics script for Next.js?
Is there any documentation or explanation why it is like that, or why it is OK to put GTM script in `<body>` tag in Next.js project, even if it's not recommended by Google?

**Comment by icyJoseph** [comment]
Asking the real questions here.

I don't have an answer, but I see that the person who updated the GTM example comes from the Google Developer Experts team that implemented the Script component for Next.js 

- https://developer.chrome.com/blog/script-component/
- https://github.com/vercel/next.js/pull/29061

That's not to say that it can't be wrong, but it just deepens the need for reconciliation between those two documents.

My understanding here, is that if you put the GTM as early as possible, other scripts can assume it exists, but maybe that does impact performance in large front end apps? So the trade off is to load it later? If so, then we can consciously make an exception to this rule, whenever needed, is that so?

I wouldn't want to tag developers out of the blue, but maybe you can open an issue, and reference #29061 and explain your concern?

**Reply by ATK3Shape** [reply]
@icyJoseph Thank you very much for the link to blog post! For some reason didn't find it before. 
Good idea about opening an issue. I will probably wait a little bit, in case someone will add new information.

**Comment by philwolstenholme** [comment, marked as answer]
Google Tag Manager wants to be loaded as soon as possible, so that more types of marketing/tracking/analytics scripts/tags can be supported, and they will do their work quicker.

But, as developers, we want Google Tag Manager to run _later_, so that it won't compete with our own scripts and network requests. Users don't care if Hotjar or Google Analytics loads late, but they will care if the page content loads slower. The downside to deferring the loading of GTM is that some people could bounce off your site before Google Analytics loads, so your bounce rate in GA wouldn't be accurate. But again, your users don't care about your GA bounce rate!

Best to use the Next Script way of doing it, I think.

**Reply by ATK3Shape** [reply]
Yes, I understand this. However, marketing team requires analytics scripts at the top of <head> tag, and don't understand, why it was always possible and is not with next.js.
I didn't find official recommendations and explanations regarding this, so I can share it with marketing team.

**Reply by icyJoseph** [reply]
Might be worth noting that the team behind the Script Tag, has a proposal to make these built-ins: https://github.com/vercel/next.js/discussions/48256

It could be good to chime in there.

**Reply by housseindjirdeh** [reply]
This is correct. @ATK3Shape: if you want to load GTM as soon as possible similar to the original guidance - you can always use the `beforeInteractive` strategy to do so. However, using `afterInteractive` is the best approach if you would not like GTM scripts to compete with critical resources on your page.

**Reply by thgh** [reply]
For optimal performance and more accurate timings, you can split the snippet in two:
dataLayer setup [ASAP](https://github.com/thgh/next-gtm/blob/main/src/index.tsx#L17-L22) using `<script>` 
gtm.js [lazy](https://github.com/thgh/next-gtm/blob/main/src/index.tsx#L37-L43) using `<Script strategy=...>` 


[Check out @thgh/next-gtm on NPM](https://www.npmjs.com/package/@thgh/next-gtm)

**Comment by housseindjirdeh** [comment]
Thanks for starting this thread. Like others mentioned, we're currently working on a library to include built-in components to simplify the developer and user experience of loading third-parties (see https://github.com/vercel/next.js/discussions/48256). Google Tag Manager is a little trickier than other third-parties, but the good news is that we're actively thinking about this. I'll likely publish a separate RFC for it soon, so stay tuned!

**Comment by lucashfreitas** [comment]
Thanks @housseindjirdeh  - on web development -  it should be straightforward to add a `<script>` tag, after the head, or before the head closing, likewise with the body.  Seems like `next/script` is opinionated and doesn't give us full control to inject code in the `head` or in `body` as we need (_which should be a very trivial task ) - glad to hear that something is being thought to solve this.

Thanks
